pkg/storegateway/stats.go

```diff
@@ -98,19 +98,40 @@ func (s queryStats) merge(o *queryStats) *queryStats {
 	return &s
 }
 
-// safeQueryStats wraps queryStats adding functions to lock/unlock a mutex before manipulating the statistics.
+// safeQueryStats wraps queryStats adding functions manipulate the statistics while holding a lock.
 type safeQueryStats struct {
-	sync.Mutex
-	*queryStats
+	unsafeStatsMx sync.Mutex
+	unsafeStats   *queryStats
 }
 
 func newSafeQueryStats() *safeQueryStats {
 	return &safeQueryStats{
-		queryStats: &queryStats{},
+		unsafeStats: &queryStats{},
 	}
 }
 
+// update the statistics while holding the lock.
+func (s *safeQueryStats) update(fn func(stats *queryStats)) {
+	s.unsafeStatsMx.Lock()
+	defer s.unsafeStatsMx.Unlock()
+
+	fn(s.unsafeStats)
+}
+
+// merge the statistics while holding the lock.
 func (s *safeQueryStats) merge(o *queryStats) *safeQueryStats {
-	s.queryStats.merge(o)
+	s.unsafeStatsMx.Lock()
+	defer s.unsafeStatsMx.Unlock()
+
+	s.unsafeStats.merge(o)
 	return s
```

Comment:
`merge` doesn't modify inplace, but returns a new pointer 

I think it would make sense to mimic this behaviour:


```suggestion
	return &safeQueryStats{unsafeStats: s.unsafeStats.merge(o)}
```

alternatively, we can modify the safe stats inplace

```suggestion
	s.unsafeStats = s.unsafeStats.merge(o)
	return s
```

Reply:
You spot a terrible bug I introduced! Thanks for noticing it. See the fix here: https://github.com/grafana/mimir/pull/3399/commits/0b2ac1b11a1f42589289ace9d59f7b4378e7e8f6

I pick the option to return a new object with the merged stats, so that we keep consistency with `queryStats.merge()`.